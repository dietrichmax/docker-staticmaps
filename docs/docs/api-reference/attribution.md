---
title: Attribution
description: ""
authors:
  - name: Max Dietrich
  - url: https://mxd.codes
last_update:
  date: 10/07/2025
  author: Max Dietrich
keywords:
  - Attribution
slug: /api-reference/attribution
sidebar_position: 3
---

# Attribution

The map service supports an optional `attribution` query parameter to control whether and how attribution is rendered on the generated image.

## Basic Behavior

- Attribution is **shown by default** (if nothing is passed).
- If no custom text is provided, the service uses the default attribution for the selected basemap.

## Parameter Format

The `attribution` parameter accepts key-value pairs separated by `|`, for example:

```
&attribution=show:true|text:© OpenStreetMap contributors
```

## Supported Keys

| Key | Description | Example |
| --- | --- | --- |
| show | `true` or `false` – controls visibility | `show:false` |
| text | Custom attribution text (URL-encoded) | `text:%C2%A9%20Your%20Company` |