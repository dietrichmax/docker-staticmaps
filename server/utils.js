import StaticMaps from "../staticmaps/staticmaps.js"
import basemaps from "./basemaps.js"

export async function render(options) {
  const map = new StaticMaps({
    width: options.width,
    height: options.height,
    tileLayers: [
      {
        tileUrl: options.tileUrl,
      },
    ],
    tileSize: 256,
  })

  let finalZoom = options.zoom
  let center = options.center

  // Handle coordinates if provided
  if (options.coordinates && options.coordinates.length > 0) {
    // Add markers if markers=true
    if (options.markers === true) {
      options.coordinates.forEach(coord => {
        map.addMarker({
          coord: coord,
          img: './public/images/marker-28.png',
          width: 42,
          height: 42,
          offsetX: 14,
          offsetY: 42
        })
      })
    }

    // Add polyline only if polyline=true and we have multiple coordinates
    if (options.polyline === true && options.coordinates.length > 1) {
      map.addLine({
        coords: options.coordinates,
        color: 'blue',
        width: 3
      })
    }

    // Calculate center and zoom only if not provided
    const bounds = getBoundingBox(options.coordinates)

    if (!center) {
      center = [(bounds.minLon + bounds.maxLon) / 2, (bounds.minLat + bounds.maxLat) / 2]
    }

    if (finalZoom === null) {
      finalZoom = calculateZoom(bounds, options.width, options.height)
    }
  }

  await map.render(center, finalZoom)
  return map.image.buffer(`image/${options.format}`, { quality: 75 })
}

export function getTileUrl(reqTileUrl, reqBasemap) {
  let tileUrl
  if (reqTileUrl) {
    tileUrl = reqTileUrl
  } else if (reqBasemap) {
    const tileUrlObj = basemaps.filter((obj) => {
      return obj.basemap === reqBasemap
    })
    return tileUrlObj[0].url
  } else {
    tileUrl = "https://tile.openstreetmap.org/{z}/{x}/{y}.png"
  }
  return tileUrl
}

export function checkParams(req) {
  const missingParams = []
  let zoom = null  // Changed to null initially

  if (req.query.zoom) {
    zoom = parseInt(req.query.zoom)
  } else if (!req.query.coordinates) {
    // Only require zoom if coordinates are not provided
    missingParams.push(" {zoom}")
  }

  // Only require center if coordinates not provided
  if (!req.query.center && !req.query.coordinates) {
    missingParams.push(" {center or coordinates}")
  }

  const markers = req.query.markers === 'true'
  const polyline = req.query.polyline === 'true'

  return { missingParams, zoom, markers, polyline }
}

export function parseCoordinates(coordString) {
  if (!coordString) return []

  return coordString.split('|').map(coord => {
    const [lat, lon] = coord.split(',').map(Number)
    return [lon, lat] // Convert to [longitude, latitude] format
  })
}

function getBoundingBox(coordinates) {
  const bounds = {
    minLat: Infinity,
    maxLat: -Infinity,
    minLon: Infinity,
    maxLon: -Infinity
  }

  coordinates.forEach(([lon, lat]) => {
    bounds.minLat = Math.min(bounds.minLat, lat)
    bounds.maxLat = Math.max(bounds.maxLat, lat)
    bounds.minLon = Math.min(bounds.minLon, lon)
    bounds.maxLon = Math.max(bounds.maxLon, lon)
  })

  return bounds
}

function calculateZoom(bounds, width, height) {
  const WORLD_PX = 256
  const ZOOM_MAX = 18
  const PADDING = 20 // 20px padding on each side

  const latRadian = bounds.maxLat * Math.PI / 180

  // Adjust width and height to account for padding
  const effectiveWidth = width - (2 * PADDING)
  const effectiveHeight = height - (2 * PADDING)

  // Calculate the width and height in pixels for the given coordinates
  const latDiff = Math.abs(bounds.maxLat - bounds.minLat)
  const lonDiff = Math.abs(bounds.maxLon - bounds.minLon)

  const latZoom = Math.floor(Math.log2(effectiveHeight * 360 / (latDiff * WORLD_PX)))
  const lonZoom = Math.floor(Math.log2(effectiveWidth * 360 / (lonDiff * WORLD_PX * Math.cos(latRadian))))

  // Use the more conservative (smaller) zoom level to ensure all points are visible
  return Math.min(Math.min(latZoom, lonZoom) - 1, ZOOM_MAX)
}