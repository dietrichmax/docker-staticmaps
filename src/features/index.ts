// src/staticmaps/shapes.ts
export { addMarkers } from "./addMarkers"
export { addPolylines } from "./addPolylines"
export { addCircles } from "./addCircles"
export { addTexts } from "./addTexts"
export { asArray } from "./asArray"
