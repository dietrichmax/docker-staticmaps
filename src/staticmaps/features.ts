// src/staticmaps/shapes.ts
export { default as IconMarker } from "./features/marker"
export { default as Polyline } from "./features/polyline"
export { default as MultiPolygon } from "./features/multipolygon"
export { default as Circle } from "./features/circle"
export { default as Text } from "./features/text"
export { default as Bound } from "./bound"
export { default as Image } from "./image"
