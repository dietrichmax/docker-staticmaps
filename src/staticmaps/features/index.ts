// src/staticmaps/features/index.ts
export { default as IconMarker } from "./marker"
export { default as Polyline } from "./polyline"
export { default as Circle } from "./circle"
export { default as Text } from "./text"
export { default as Bound } from "../bound"
export { default as Image } from "../image"
