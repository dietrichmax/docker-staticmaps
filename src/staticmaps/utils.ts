// utils.ts
import { Coordinate } from "src/types/types"
import logger from "../utils/logger"

/**
 * Transform longitude to tile number.
 *
 * @param lon - Longitude value.
 * @param zoom - Zoom level.
 * @returns The tile number corresponding to the given longitude at the specified zoom level.
 */
export function lonToX(lon: number, zoom: number): number {
  if (!(lon >= -180 && lon <= 180)) {
    lon = ((((lon + 180) % 360) + 360) % 360) - 180
  }
  return ((lon + 180) / 360) * Math.pow(2, zoom)
}

/**
 * Transform latitude to tile number.
 *
 * @param lat - Latitude value.
 * @param zoom - Zoom level.
 * @returns The tile number corresponding to the given latitude at the specified zoom level.
 */
export function latToY(lat: number, zoom: number): number {
  const MAX_LATITUDE = 85.05112878
  if (lat > MAX_LATITUDE) lat = MAX_LATITUDE
  if (lat < -MAX_LATITUDE) lat = -MAX_LATITUDE

  return (
    ((1 -
      Math.log(
        Math.tan((lat * Math.PI) / 180) + 1 / Math.cos((lat * Math.PI) / 180)
      ) /
        Math.PI) /
      2) *
    Math.pow(2, zoom)
  )
}

/**
 * Transform tile number to latitude.
 *
 * @param y - Tile number.
 * @param zoom - Zoom level.
 * @returns The latitude corresponding to the tile number at the specified zoom level.
 */
export function yToLat(y: number, zoom: number): number {
  return (
    (Math.atan(Math.sinh(Math.PI * (1 - (2 * y) / Math.pow(2, zoom)))) /
      Math.PI) *
    180
  )
}

/**
 * Transform tile number to longitude.
 *
 * @param x - Tile number.
 * @param zoom - Zoom level.
 * @returns The longitude corresponding to the tile number at the specified zoom level.
 */
export function xToLon(x: number, zoom: number): number {
  return (x / Math.pow(2, zoom)) * 360.0 - 180.0
}

/**
 * Convert meters to pixels at a given zoom level and latitude.
 *a
 * @param meter - Distance in meters.
 * @param zoom - Zoom level.
 * @param lat - Latitude at which to calculate the conversion.
 * @returns The equivalent pixel distance.
 */
export const meterToPixel = (
  meter: number,
  zoom: number,
  lat: number
): number => {
  const latitudeRadians = lat * (Math.PI / 180)
  const meterPerPixel = (156543.03392 * Math.cos(latitudeRadians)) / 2 ** zoom
  return meter / meterPerPixel
}

/**
 * Process a queue of async functions sequentially.
 *
 * @param queue - Array of async functions.
 * @param index - Current index (default is 0).
 * @returns A promise that resolves when the queue is processed.
 */
export const workOnQueue = async <T>(
  queue: (() => Promise<T>)[],
  index = 0
): Promise<boolean> => {
  if (!queue[index]) return true
  await queue[index]()
  return workOnQueue(queue, index + 1)
}

/**
 * Split an array into chunks of a specified size.
 *
 * @param array - The array to split.
 * @param size - The chunk size.
 * @returns A new array containing chunks.
 */
export function chunk<T>(array: T[], size: number): T[][] {
  const result: T[][] = []
  for (let i = 0; i < array.length; i += size) {
    result.push(array.slice(i, i + size))
  }
  return result
}

/**
 * Converts tile coordinates (x, y, z) to a QuadKey.
 *
 * @param {number} x - The x coordinate of the tile.
 * @param {number} y - The y coordinate of the tile.
 * @param {number} z - The zoom level of the tile.
 * @returns {string} - The corresponding QuadKey as a string.
 */
export function tileXYToQuadKey(x: number, y: number, z: number): string {
  const quadKey: string[] = []
  for (let i = z; i > 0; i--) {
    let digit = "0"
    const mask = 1 << (i - 1)
    if ((x & mask) !== 0) digit = (parseInt(digit) + 1).toString()
    if ((y & mask) !== 0) {
      digit = (parseInt(digit) + 2).toString()
    }
    quadKey.push(digit)
  }
  return quadKey.join("")
}

/**
 * Generate a geodesic line between two coordinates.
 * @param start - [lat, lon]
 * @param end - [lat, lon]
 * @param segments - Number of interpolation points
 * @returns Array of [lon, lat] points (for GeoJSON, static maps, etc.)
 */
export function createGeodesicLine(
  start: Coordinate,
  end: Coordinate,
  segments: number = 70
): Coordinate[] {
  const toRadians = (deg: number): number => (deg * Math.PI) / 180
  const toDegrees = (rad: number): number => (rad * 180) / Math.PI

  const lat1 = toRadians(start[0])
  const lon1 = toRadians(start[1])
  const lat2 = toRadians(end[0])
  const lon2 = toRadians(end[1])

  const delta = Math.acos(
    Math.sin(lat1) * Math.sin(lat2) +
      Math.cos(lat1) * Math.cos(lat2) * Math.cos(lon2 - lon1)
  )

  if (delta === 0)
    return [
      [start[1], start[0]],
      [end[1], end[0]],
    ] // [lon, lat]

  const geodesic: Coordinate[] = []

  for (let i = 0; i <= segments; i++) {
    const f = i / segments
    const A = Math.sin((1 - f) * delta) / Math.sin(delta)
    const B = Math.sin(f * delta) / Math.sin(delta)

    const x =
      A * Math.cos(lat1) * Math.cos(lon1) + B * Math.cos(lat2) * Math.cos(lon2)
    const y =
      A * Math.cos(lat1) * Math.sin(lon1) + B * Math.cos(lat2) * Math.sin(lon2)
    const z = A * Math.sin(lat1) + B * Math.sin(lat2)

    const lat = Math.atan2(z, Math.sqrt(x * x + y * y))
    const lon = Math.atan2(y, x)

    geodesic.push([toDegrees(lon), toDegrees(lat)]) // [lon, lat]
  }

  return geodesic
}

/**
 * Smooths a series of coordinates using the Chaikin's algorithm.
 *
 * @param {Array<Coordinate>} coords The input coordinates to smooth.
 * @param {number} [iterations=2] The number of iterations for smoothing.
 *
 * @returns {Array<Coordinate>} The smoothed set of coordinates.
 */
export function chaikinSmooth(
  coords: Coordinate[],
  iterations: number = 2
): Coordinate[] {
  if (coords.length < 2) return coords // early return for empty or single point

  for (let i = 0; i < iterations; i++) {
    const newCoords: Coordinate[] = []
    for (let j = 0; j < coords.length - 1; j++) {
      const [x0, y0] = coords[j]
      const [x1, y1] = coords[j + 1]
      const q: Coordinate = [0.75 * x0 + 0.25 * x1, 0.75 * y0 + 0.25 * y1]
      const r: Coordinate = [0.25 * x0 + 0.75 * x1, 0.25 * y0 + 0.75 * y1]
      newCoords.push(q, r)
    }
    coords = [coords[0], ...newCoords, coords[coords.length - 1]]
  }
  return coords
}

/**
 * Simplifies a series of coordinates using the Douglas-Peucker algorithm.
 *
 * @param {Array<Coordinate>} coords The input coordinates to simplify.
 * @param {number} epsilon The maximum distance allowed between original and simplified points.
 *
 * @returns {Array<Coordinate>} The simplified set of coordinates.
 */
/**
 * Simplifies a series of coordinates using the Douglas-Peucker algorithm.
 *
 * @param {Array<Coordinate>} coords The input coordinates to simplify.
 * @param {number} epsilon The maximum distance allowed between original and simplified points.
 *
 * @returns {Array<Coordinate>} The simplified set of coordinates.
 */
export function douglasPeucker(
  coords: Coordinate[],
  epsilon: number
): Coordinate[] {
  const dmax = (
    points: Coordinate[],
    from: number,
    to: number
  ): { index: number; dist: number } => {
    let maxDist = 0
    let index = from
    const [x1, y1] = points[from]
    const [x2, y2] = points[to]
    const dx = x2 - x1
    const dy = y2 - y1
    const len2 = dx * dx + dy * dy

    for (let i = from + 1; i < to; i++) {
      const [x0, y0] = points[i]
      const t = len2 === 0 ? 0 : ((x0 - x1) * dx + (y0 - y1) * dy) / len2
      const px = x1 + t * dx
      const py = y1 + t * dy
      const dist = Math.sqrt((x0 - px) ** 2 + (y0 - py) ** 2)
      if (dist > maxDist) {
        index = i
        maxDist = dist
      }
    }
    return { index, dist: maxDist }
  }

  const simplify = (
    points: Coordinate[],
    from: number,
    to: number
  ): Coordinate[] => {
    const { index, dist } = dmax(points, from, to)
    if (dist > epsilon) {
      const rec1 = simplify(points, from, index)
      const rec2 = simplify(points, index, to)
      return rec1.slice(0, -1).concat(rec2)
    } else {
      return [points[from], points[to]]
    }
  }

  if (coords.length < 3) return coords
  return simplify(coords, 0, coords.length - 1)
}

/**
 * Validates and normalizes a strokeDasharray option.
 *
 * Ensures the value is an array of non-negative numbers.
 * Throws an error if the validation fails.
 *
 * @param {any} strokeDasharray - The strokeDasharray value to validate.
 * @returns {number[]} The validated strokeDasharray array.
 * @throws {Error} If strokeDasharray is not an array of non-negative numbers.
 */
export function normalizeStrokeDasharray(strokeDasharray: any): number[] {
  if (strokeDasharray === undefined) {
    return []
  }
  if (
    Array.isArray(strokeDasharray) &&
    strokeDasharray.every(
      (n) => typeof n === "number" && n >= 0 && Number.isFinite(n)
    )
  ) {
    return strokeDasharray
  }
  throw new Error(
    "Invalid strokeDasharray: must be an array of non-negative finite numbers"
  )
}
