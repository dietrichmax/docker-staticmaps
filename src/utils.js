import StaticMaps from "./staticmaps/staticmaps.js"
import basemaps from "./basemaps.js"

export function validateParams(req) {
  const missingParams = [] // to render feedback for missing params
  let center = null // Changed to null initially
  let polyline = false
  let markers = false

  if (req.query.center) {
    const coordinates = req.query.center.split(",")
    center = [parseFloat(coordinates[0]), parseFloat(coordinates[1])]
  }

  if (req.query.polyline) {
    let polylineWeight = 5 // default polyline weight
    let polylineColor = "blue" // default polyline color
    let removeValFromIndex = []

    let pathArray = req.query.polyline.split("|")
    let polylineCoords = pathArray
    pathArray.map((elem) => {
      if (elem.includes("color")) {
        //--> todo validate color function hex and name
        // get color param for polyline and the index of it
        removeValFromIndex.push(pathArray.indexOf(elem))
        polylineColor = elem.replace("color:", "#")
      }

      if (elem.includes("weight")) {
        // get weight param for polyline and the index of it
        removeValFromIndex.push(pathArray.indexOf(elem))
        polylineWeight = parseInt(elem.replace("weight:", ""))
      }
    })

    // remove color and weight from polyline query so coordinates are left
    if (removeValFromIndex.length > 0) {
      for (var i = removeValFromIndex.length - 1; i >= 0; i--) {
        polylineCoords.splice(removeValFromIndex[i], 1)
      }
    }

    // sort coordinates
    polylineCoords = Object.keys(pathArray).map((key) => {
      const [lat, lon] = pathArray[key].split(",").map(Number)
      return [lon, lat] // Convert to [longitude, latitude] format
    })

    polyline = {
      coords: polylineCoords,
      weight: polylineWeight,
      color: polylineColor,
    }
  }

  if (req.query.markers) {
    let markersArray = req.query.markers.split("|")
    let markersCoords = markersArray
    markersCoords = Object.keys(markersArray).map((key) => {
      const [lat, lon] = markersCoords[key].split(",").map(Number)
      return [lon, lat] // Convert to [longitude, latitude] format
    })

    markers = {
      coords: markersCoords,
    }
  }
  // Only require center if polyline or marker coordinates not provided
  if (!req.query.center && !polyline.coords && !markers.coords) {
    missingParams.push(" {center or coordinates}")
  }

  const options = {
    width: parseInt(req.query.width) || 300,
    height: parseInt(req.query.height) || 300,
    zoom: parseInt(req.query.zoom),
    center: center,
    markers: markers,
    polyline: polyline,
    tileUrl: getTileUrl(req.query.tileUrl, req.query.basemap),
    format: req.query.format || "png",
  }
  return { missingParams, options }
}

export async function render(options) {
  let finalZoom = options.zoom
  let center = options.center

  const map = new StaticMaps({
    width: options.width,
    height: options.height,
    tileLayers: [
      {
        tileUrl: options.tileUrl,
      },
    ],
    tileSize: 256,
  })

  // Add markers only if markers object exists and we atleast one coordinates
  if (options.markers && options.markers.coords.length > 0) {
    options.markers.coords.forEach((coord) => {
      map.addMarker({
        coord: coord,
        img: "./public/images/marker-28.png",
        width: 42,
        height: 42,
        offsetX: 15,
        offsetY: 27,
      })
    })
  }

  // Add polyline only if polyline object exists and we have multiple coordinates
  if (options.polyline && options.polyline.coords.length > 1) {
    map.addLine({
      coords: options.polyline.coords,
      color: options.polyline.color,
      width: options.polyline.weight,
    })

    // Calculate polyline center and zoom only if not provided
    const bounds = getBoundingBox(options.polyline.coords)

    if (!center) {
      center = [
        (bounds.minLon + bounds.maxLon) / 2,
        (bounds.minLat + bounds.maxLat) / 2,
      ]
    }
  }

  if (finalZoom === null) {
    finalZoom = calculateZoom(bounds, options.width, options.height)
  }
  //}

  await map.render(center, finalZoom)
  return map.image.buffer(`image/${options.format}`, { quality: 75 })
}

export function getTileUrl(reqTileUrl, reqBasemap) {
  let tileUrl
  if (reqTileUrl) {
    tileUrl = reqTileUrl
  } else if (reqBasemap) {
    const tileUrlObj = basemaps.filter((obj) => {
      return obj.basemap === reqBasemap
    })
    return tileUrlObj[0].url
  } else {
    tileUrl = "https://tile.openstreetmap.org/{z}/{x}/{y}.png"
  }
  return tileUrl
}

export function parseCoordinates(coordString) {
  if (!coordString) return []

  return coordString.split("|").map((coord) => {
    const [lat, lon] = coord.split(",").map(Number)
    return [lon, lat] // Convert to [longitude, latitude] format
  })
}

function getBoundingBox(coordinates) {
  const bounds = {
    minLat: Infinity,
    maxLat: -Infinity,
    minLon: Infinity,
    maxLon: -Infinity,
  }

  coordinates.forEach(([lon, lat]) => {
    bounds.minLat = Math.min(bounds.minLat, lat)
    bounds.maxLat = Math.max(bounds.maxLat, lat)
    bounds.minLon = Math.min(bounds.minLon, lon)
    bounds.maxLon = Math.max(bounds.maxLon, lon)
  })

  return bounds
}

function calculateZoom(bounds, width, height) {
  const WORLD_PX = 256
  const ZOOM_MAX = 18
  const PADDING = 20 // 20px padding on each side

  const latRadian = (bounds.maxLat * Math.PI) / 180

  // Adjust width and height to account for padding
  const effectiveWidth = width - 2 * PADDING
  const effectiveHeight = height - 2 * PADDING

  // Calculate the width and height in pixels for the given coordinates
  const latDiff = Math.abs(bounds.maxLat - bounds.minLat)
  const lonDiff = Math.abs(bounds.maxLon - bounds.minLon)

  const latZoom = Math.floor(
    Math.log2((effectiveHeight * 360) / (latDiff * WORLD_PX))
  )
  const lonZoom = Math.floor(
    Math.log2(
      (effectiveWidth * 360) / (lonDiff * WORLD_PX * Math.cos(latRadian))
    )
  )

  // Use the more conservative (smaller) zoom level to ensure all points are visible
  return Math.min(Math.min(latZoom, lonZoom) - 1, ZOOM_MAX)
}
