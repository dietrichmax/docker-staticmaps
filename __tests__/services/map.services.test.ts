import {
  generateMap,
  addCircles,
  addPolylines,
  addTexts,
  addMarkers,
  asArray,
} from "../../src/services/map.services" // adjust the import path as needed
import StaticMaps from "../../src/staticmaps/staticmaps"
import { createAttributionSVG } from "../../src/utils/attribution"
import sharp from "sharp"
import logger from "../../src/utils/logger"
import { IconMarker } from "../../src/staticmaps/features"

// Mock dependencies
jest.mock("../../src/staticmaps/staticmaps")
jest.mock("../../src/utils/logger", () => ({
  debug: jest.fn(),
  info: jest.fn(),
  warn: jest.fn(),
  error: jest.fn(),
}))
jest.mock("../../src/utils/attribution", () => ({
  createAttributionSVG: jest.fn(() => Buffer.from("<svg></svg>")),
}))
jest.mock("sharp")

describe("generateMap", () => {
  let renderMock: jest.Mock
  let bufferMock: jest.Mock

  beforeEach(() => {
    renderMock = jest.fn()
    bufferMock = jest.fn(() => Promise.resolve(Buffer.from("image-buffer")))
    ;(StaticMaps as jest.Mock).mockImplementation(() => ({
      render: renderMock,
      image: {
        buffer: bufferMock,
      },
      addMarker: jest.fn(),
      addLine: jest.fn(),
      addPolygon: jest.fn(),
      addCircle: jest.fn(),
      addText: jest.fn(),
    }))
    ;(sharp as unknown as jest.Mock).mockReturnValue({
      composite: jest.fn().mockReturnThis(),
      toFormat: jest.fn().mockReturnThis(),
      toBuffer: jest.fn().mockResolvedValue(Buffer.from("composited-image")),
    })
  })

  it("renders a map with all feature types", async () => {
    const options = {
      width: 800,
      height: 600,
      format: "png",
      zoom: 10,
      center: [0, 0],
      markers: [{ coords: [[1, 2]], img: "pin.png", width: 20, height: 20 }],
      polyline: [
        {
          coords: [
            [0, 0],
            [1, 1],
          ],
          color: "#ff0000",
          weight: 2,
        },
      ],
      polygon: [
        {
          coords: [
            [0, 0],
            [1, 1],
            [1, 0],
          ],
          color: "#00ff00",
          fill: "#00ff00",
        },
      ],
      circle: [{ coords: [[1, 1]], radius: 100, color: "#0000ff", width: 1 }],
      text: [{ coords: [[2, 2]], text: "Hello", size: 12 }],
    }

    const buffer = await generateMap(options)

    expect(renderMock).toHaveBeenCalledWith(options.center, options.zoom)
    expect(buffer.equals(Buffer.from("image-buffer"))).toBe(true)
  })

  it("adds attribution overlay if enabled", async () => {
    const options = {
      width: 800,
      height: 600,
      format: "png",
      zoom: 10,
      center: [0, 0],
      attribution: {
        show: true,
        text: "Â© OpenStreetMap contributors",
      },
    }

    const buffer = await generateMap(options)

    expect(createAttributionSVG).toHaveBeenCalledWith(
      options.attribution.text,
      options.width,
      options.height
    )
    expect(buffer.equals(Buffer.from("composited-image"))).toBe(true)
  })

  /*it("does not add attribution overlay if disabled", async () => {
    const options = {
      width: 800,
      height: 600,
      format: "png",
      zoom: 10,
      center: [0, 0],
      attribution: "show:false"
    }

    const buffer = await generateMap(options)

    expect(createAttributionSVG).not.toHaveBeenCalled()
    expect(buffer.equals(Buffer.from("image-buffer"))).toBe(true)
  })*/

  it("throws if map.image is undefined", async () => {
    ;(StaticMaps as jest.Mock).mockImplementation(() => ({
      render: renderMock,
      image: undefined,
    }))

    await expect(
      generateMap({ center: [0, 0], zoom: 10, format: "png" })
    ).rejects.toThrow("Map image is undefined after rendering")
  })

  it("returns expected buffer snapshot", async () => {
    const buffer = await generateMap({
      center: [0, 0],
      zoom: 10,
      format: "png",
    })
    expect(buffer.toString()).toMatchSnapshot()
  })
})

describe("asArray", () => {
  it("returns an empty array when input is falsy", () => {
    expect(asArray(null)).toEqual([])
    expect(asArray(undefined)).toEqual([])
  })

  it("wraps non-array input into an array", () => {
    expect(asArray("marker")).toEqual(["marker"])
    expect(asArray(1)).toEqual([1])
  })

  it("returns the same array if input is already an array", () => {
    const input = [1, 2, 3]
    expect(asArray(input)).toBe(input)
  })
})

describe("addPolylines", () => {
  it("skips polyline with less than 2 coords and logs a warning", () => {
    const mockMap = { addLine: jest.fn(), addPolygon: jest.fn() }
    const item = { coords: [[1, 2]], color: "red", weight: 1 }

    addPolylines(mockMap as any, [item], false)

    expect(mockMap.addLine).not.toHaveBeenCalled()
    expect(logger.warn).toHaveBeenCalledWith(
      "Skipping polyline [0] due to insufficient coords",
      item
    )
  })

  it("skips polygon with less than 2 coords and logs a warning", () => {
    const mockMap = { addPolygon: jest.fn() }
    const item = { coords: [[1, 2]], color: "blue", fill: "#0000ff" }

    addPolylines(mockMap as any, [item], true)

    expect(mockMap.addPolygon).not.toHaveBeenCalled()
    expect(logger.warn).toHaveBeenCalledWith(
      "Skipping polygon [0] due to insufficient coords",
      item
    )
  })

  it("adds multiple polylines and polygons", () => {
    const mockMap = { addLine: jest.fn(), addPolygon: jest.fn() }
    const polylineItems = [
      {
        coords: [
          [0, 0],
          [1, 1],
        ],
        color: "red",
        weight: 1,
      },
      {
        coords: [
          [1, 1],
          [2, 2],
        ],
        color: "blue",
        weight: 2,
      },
    ]

    const polygonItems = [
      {
        coords: [
          [0, 0],
          [1, 1],
          [1, 0],
        ],
        color: "#00ff00",
        fill: "#00ff00",
      },
    ]

    addPolylines(mockMap as any, polylineItems, false)
    addPolylines(mockMap as any, polygonItems, true)

    expect(mockMap.addLine).toHaveBeenCalledTimes(2)
    expect(mockMap.addPolygon).toHaveBeenCalledTimes(1)
  })
})

describe("addCircles", () => {
  it("skips circle with no coords and logs a warning", () => {
    const mockMap = { addCircle: jest.fn() }
    const circ = { coords: [], radius: 100 }

    addCircles(mockMap as any, [circ])

    expect(mockMap.addCircle).not.toHaveBeenCalled()
    expect(logger.warn).toHaveBeenCalledWith(
      "Skipping circle [0] due to missing coords",
      circ
    )
  })
})

describe("addTexts", () => {
  it("skips text with no coords and logs a warning", () => {
    const mockMap = { addText: jest.fn() }
    const txt = { coords: [], text: "Test Label" }

    addTexts(mockMap as any, [txt])

    expect(mockMap.addText).not.toHaveBeenCalled()
    expect(logger.warn).toHaveBeenCalledWith(
      "Skipping text [0] due to missing coords",
      txt
    )
  })
})

describe("addMarkers", () => {
  const mockAddMarker = jest.fn()
  const mockMap = {
    addMarker: mockAddMarker,
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it("adds multiple markers with multiple coords each", () => {
    const markerData = [
      {
        coords: [
          [10, 20],
          [30, 40],
        ],
        img: "marker.png",
        width: 32,
        height: 32,
        offsetX: 0,
        offsetY: 0,
        color: "#f00",
      },
    ]

    addMarkers(mockMap as any, markerData)

    // Should call addMarker twice (2 coords)
    expect(mockAddMarker).toHaveBeenCalledTimes(2)

    // Check type and values
    const firstCallArg = mockAddMarker.mock.calls[0][0]
    expect(firstCallArg).toBeInstanceOf(IconMarker)
    expect(firstCallArg.coord).toEqual([10, 20])

    const secondCallArg = mockAddMarker.mock.calls[1][0]
    expect(secondCallArg.coord).toEqual([30, 40])

    // Debug logs called for both
    expect(logger.debug).toHaveBeenCalledTimes(2)
    expect(logger.debug).toHaveBeenCalledWith(
      "Adding marker [0][0]",
      markerData[0]
    )
    expect(logger.debug).toHaveBeenCalledWith(
      "Adding marker [0][1]",
      markerData[0]
    )
  })

  it("skips marker with empty coords array", () => {
    const markerData = [{ coords: [] }]
    addMarkers(mockMap as any, markerData)

    expect(mockAddMarker).not.toHaveBeenCalled()
    expect(logger.debug).not.toHaveBeenCalled()
  })

  it("does nothing if no markers provided", () => {
    addMarkers(mockMap as any, [])

    expect(mockAddMarker).not.toHaveBeenCalled()
    expect(logger.debug).not.toHaveBeenCalled()
  })
})
