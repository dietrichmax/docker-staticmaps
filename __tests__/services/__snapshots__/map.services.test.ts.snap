// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateMap returns expected buffer snapshot 1`] = `"image-buffer"`;
